### DISCUSSION TITLE
Stopping an asynchronous server

### DISCUSSION THREAD

**Discussion by amigian74** [pymodbus-dev/pymodbus#2485] [Help] [answered]
Hi there,

I know a similiar question was asked here before and answered but the answer won't work for me at all. I have a thread class which starts an asynchronous modbus server before run. Within a loop this class gets data which is populated to the modbus server. Until this point everything works fine. At some time the thread is canceled but I didÄnt find any way to stop the modbus server as well. Here is a slimmed down example:

```
class Plugin(threading.Thread)

async def _run_server(self):
  server = await StartAsyncTcpServer(...)
  return server

def _start_server(self):
  asyncio.run(self._run_server())

def initialize(self):
  self._server_thread = threading.Thread(target=self._start_server)
  self._server_thread.start()

def run(self):
  while True:
    # get data and populate it
    if self._stop_event.is_set():
      break
  asyncio.run(ServerAsyncStop())
  self._server_thread.join()
```

Hope anybody can help me out here. Thx in advance Steven

**Comment by janiversen** [comment, marked as answer]
You need to use ModbusAsyncTcpServer instead and build your own logic around it.

You are effectively creating 2 async loops (2 asyncio.run) so they cannot see each other.

**Reply by amigian74** [reply]
Hi-

Sry, but this isn't an answer. The pymodbus package provides StartAsyncTcpServer and ServerAsyncStop and there is no working example for it? I've tried other solutions which didÄnt work either. Could you please provide a small example with another approach.

Thx Steven

**Reply by janiversen** [reply]
It is the answer sorry about that. You mix sync and async in a way that cannot work (open 2 async loops). stop() works when called with the same loop.

The test harness does exactly that....starts a server, runs the test and stop the server.

As I wrote earlier the start/stop functions are provided to make simple servers, having 2 async loop is far beyond that.

**Reply by janiversen** [reply]
If you do not intentionally want to have 2 async loops, you should look for general asyncio examples on how to avoid that, that is not a pymodbus problem.

**Comment by janiversen** [comment]
Let me try to analyze your program, to show why we cannot provide examples

You use Thread so your program is clearly a sync. program.

You call startAsyncModbusServer instead of the sync. variant so it's fair to assume you need a part of your program to be async.

You then start a second loop with asyncio.run to call serverAsyncStop instead of using the sync version.

Either you have a very advanced program (2 asyncio loop, is something you do not see every day.
Or you have not read the pymodbus documentation (server api part)
Or you have missed a bit about how asyncio works.

Anyhow I hope my explanation shows you why we cannot provide examples for such advanced usage, and to be honest for advanced usage it is a lot to use the ModbusAsyncServer directly.
